lib/Transport.js

```diff
@@ -347,91 +356,124 @@ class Transport {
+            params.body = buffer
+            makeRequest()
+          })
+        } else {
+          params.headers['content-length'] = '' + Buffer.byteLength(params.body)
+          makeRequest()
         }
       } else {
-        params.body = params.bulkBody
-      }
-      if (params.body !== '') {
-        headers['content-type'] = headers['content-type'] || 'application/x-ndjson'
+        makeRequest()
       }
     }
 
-    params.headers = headers
-    // serializes the querystring
-    if (options.querystring == null) {
-      params.querystring = this.serializer.qserialize(params.querystring)
-    } else {
-      params.querystring = this.serializer.qserialize(
-        Object.assign({}, params.querystring, options.querystring)
-      )
-    }
-
-    // handles request timeout
-    params.timeout = toMs(options.requestTimeout || this.requestTimeout)
-    if (options.asStream === true) params.asStream = true
-    meta.request.params = params
-    meta.request.options = options
-
-    // handle compression
-    if (params.body !== '' && params.body != null) {
-      if (isStream(params.body) === true) {
-        if (compression === 'gzip') {
-          params.headers['content-encoding'] = compression
-          params.body = params.body.pipe(createGzip())
-        }
-        makeRequest()
-      } else if (compression === 'gzip') {
-        gzip(params.body, (err, buffer) => {
-          /* istanbul ignore next */
-          if (err) {
+    // still need to check the product or waiting for the check to finish
+    if (this[kProductCheck] === 0 || this[kProductCheck] === 1) {
+      // let pass info requests
+      if (params.method === 'GET' && params.path === '/') {
+        prepareRequest()
+      } else {
+        // wait for product check to finish
+        productCheckEmitter.once('product-check', status => {
+          if (status === false) {
+            const err = new ElasticsearchClientError('The client noticed that the server is not Elasticsearch and we do not support this unknown product.')
             this.emit('request', err, result)
-            return callback(err, result)
+            process.nextTick(callback, err, result)
+          } else {
+            prepareRequest()
           }
-          params.headers['content-encoding'] = compression
-          params.headers['content-length'] = '' + Buffer.byteLength(buffer)
-          params.body = buffer
-          makeRequest()
         })
-      } else {
-        params.headers['content-length'] = '' + Buffer.byteLength(params.body)
-        makeRequest()
+        // the very first request triggers the product check
+        if (this[kProductCheck] === 0) {
+          this.productCheck()
+        }
       }
+    // the product check is finished and it's not Elasticsearch
+    } else if (this[kProductCheck] === 3) {
+      const err = new ElasticsearchClientError('The client noticed that the server is not Elasticsearch and we do not support this unknown product.')
```

Comment:
Maybe it makes sense to have a dedicated error class for this case? It would allow Kibana to distinguish between this critical error and another `ElasticsearchClientError`

Reply:
Agreed!